doc/arch/adr-010-learn-to-rank.md

```diff
@@ -0,0 +1,286 @@
+
+In a [validation spike][] we found that we could build a model that provided
+a better nDCG score than our current search algorithm.
+
+This validated for us that Learning To Rank is feasible and could
+lead to more relevant results for users.
+
+Though Learning To Rank won't make relevance tuning easier, it could
+make search results more relevant.
+
+### Final decision
+
+We feel confident about implementing Learning To Rank (LTR) for the
+following reasons:
+
+- We think that it will improve search relevancy for citizens, enabling
+  them to find what they need on GOV.UK
+- We have validated our assumptions with a spike
+- LTR is a well trodden path, and other organisations have written about
+  their implementations with varying levels of openness (e.g. Wikimedia)
+- GOV.UK has implemented other machine learning projects with success
+  (related links, topic taxonomy), and we have a data science team which
+  the search team has already worked with
+- TensorFlow provides an off-the-shelf ranking module, which solves
+  some of the machine learning problems of LTR for us
+- Manual relevancy tuning is not scalable, and won't get us to the same
+  level of relevancy as LTR
+
+## Implementation plan
+
+We have decided to implement Learning To Rank by following on from
+the work started in the [validation spike][].
+
+As we are at the beginning of the project, this section may not
+accurately reflect the final implementation.
+
+[validation spike]: https://github.com/alphagov/search-api/pull/1768
+
+### Language
+
+We would like to keep the code for this project in Ruby as much as
+possible. Search API is written in Ruby and it is the primary language
+used at GOV.UK. Necessarily some machine learning code will need to be
+written in Python.
+
+### Machine learning library
+
+We have decided to [TensorFlow's ranking library][] for our
+implementation of Learning To Rank.
+
+TensorFlow is well supported, and the ranking library is under active development.
+
+We considered using the [elasticsearch-learning-to-rank][] plugin.
+However, AWS Elasticsearch does not permit us to install custom plugins.
+
+We also considered PyTorch. However, this would require writing much more
+custom machine learning code, as PyTorch don't currently have a ranking library.
+
+[TensorFlow's ranking library]: https://github.com/tensorflow/ranking
+[elasticsearch-learning-to-rank]: https://github.com/o19s/elasticsearch-learning-to-rank
+
+### Dockerised model
+
+We will use TensorFlow Serving to make a re-ranker model available
+to Search API on search machines.
+
+We will run a docker container (containing our pre-trained model) on
+search machines to which Search API can send requests.
+
+Should the docker container fail, Search API should be resilient to this,
+and fall back to the default ranking algorithm, with results not
+re-ranked by the model.
+
+### Training
+
+**Training the model** should happen on search machines on a schedule.
```

Comment:
> we fall back to non-LTR enhanced search until we've generated a new one?

Yes I thought this would work. It'll be degraded, but still will return results. Generating a new model should be straightforward provided we have backups of generated relevancy judgements / training data (or a way to regenerate those, too).

My thinking was that one machine would generate the model file, and upload it to S3. Other machines would then pull the latest version of the model from S3.

If we want additional redundancy we could backup the S3 bucket.

With the models in the S3 bucket we can have model versions available to pull without waiting for re-training, and could roll back to different models quickly.

TensorFlow Serving can serve multiple models from one container. They also have some versioning built in for switching over to new model versions (https://www.tensorflow.org/tfx/serving/architecture#version_policy).

Reply:
Do we have and do we need the ability to fallback to the non-LTR enhanced search at any point should we need to?

Reply:
In terms of versioning, do we also need to think about how iterating the algorithm will impact users and whether this is subject to further testing?

Reply:
☝️ What @karlbaker02 said! Do you need to be able to run multiple models in production at the same time, to do multivariate testing?

Reply:
We can fall back to the non-LTR search. Perhaps we could keep a feature flag on the reranker for debugging and rollback purposes.

Yes, I think we would need multiple models in production at the same time to do an AB test.